### PR TITLE
Introduce tree visiting hierarchy printer

### PR DESCRIPTION
**Pull request by kunli2**
Adds a visitor to print LST visiting order and hierarchy.
This visitor can help new developers (including me) to understand what the LST looks like, and what `visitX` method should use when implementing a recipe.

Skip unvisited elements like `JRightPadded` or `JLeftPadded` from the output, but keeps an option to print all elements which may be extensible in the future. 

An example:
input source code :
```
              import java.io.BufferedReader;
              class T {
                  public void doSomething(StringBuilder sb, BufferedReader br) {
                      String line;
                      try {
                          while ((line = br.readLine()) != null) {
                              sb.append(line);
                          }
                      } catch (Exception e) {
                          error("Exception", e);
                      }
                  }
                  private static void error(String s, Exception e) {
                  
                  }
              }
```

An example usage, add this code in any recipe visitor to print the LST to the console.
```
            @Override
            public J.CompilationUnit visitCompilationUnit(J.CompilationUnit cu, ExecutionContext executionContext) {
                System.out.println("------visitCompilationUnit LST Tree Start------");
                System.out.println(TreeVisitingPrinter.printTree(cu));
                System.out.println("------visitCompilationUnit LST Tree End------");
                return super.visitCompilationUnit(cu, executionContext);
            }
```

output
```
------visitCompilationUnit LST Tree Start------
----J.CompilationUnit
    |-------J.Import | "import java.io.BufferedReader"
    |       \---J.FieldAccess | "java.io.BufferedReader"
    |           |---J.FieldAccess | "java.io"
    |           |   |---J.Identifier | "java"
    |           |   \-------J.Identifier | "io"
    |           \-------J.Identifier | "BufferedReader"
    \---J.ClassDeclaration
        |---J.Identifier | "T"
        \---J.Block
            |-------J.MethodDeclaration | "MethodDeclaration{T{name=doSomething,return=void,parameters=[java.lang.StringBuilder,java.io.BufferedReader]}}"
            |       |---J.Modifier | "public"
            |       |---J.Primitive | "void"
            |       |---J.Identifier | "doSomething"
            |       |-----------J.VariableDeclarations | "StringBuilder sb"
            |       |   |       |---J.Identifier | "StringBuilder"
            |       |   |       \-------J.VariableDeclarations.NamedVariable | "sb"
            |       |   |               \---J.Identifier | "sb"
            |       |   \-------J.VariableDeclarations | "BufferedReader br"
            |       |           |---J.Identifier | "BufferedReader"
            |       |           \-------J.VariableDeclarations.NamedVariable | "br"
            |       |                   \---J.Identifier | "br"
            |       \---J.Block
            |           |-------J.VariableDeclarations | "String line"
            |           |       |---J.Identifier | "String"
            |           |       \-------J.VariableDeclarations.NamedVariable | "line"
            |           |               \---J.Identifier | "line"
            |           \-------J.Try
            |                   |---J.Block
            |                   |   \-------J.WhileLoop
            |                   |           |---J.ControlParentheses | "((line = br.readLine()) != null)"
            |                   |           |   \-------J.Binary | "(line = br.readLine()) != null"
            |                   |           |           |---J.Parentheses | "(line = br.readLine())"
            |                   |           |           |   \-------J.Assignment | "line = br.readLine()"
            |                   |           |           |           |---J.Identifier | "line"
            |                   |           |           |           \-------J.MethodInvocation | "br.readLine()"
            |                   |           |           |                   |-------J.Identifier | "br"
            |                   |           |           |                   |---J.Identifier | "readLine"
            |                   |           |           |                   \-----------J.Empty
            |                   |           |           \---J.Literal
            |                   |           \-------J.Block
            |                   |                   \-------J.MethodInvocation | "sb.append(line)"
            |                   |                           |-------J.Identifier | "sb"
            |                   |                           |---J.Identifier | "append"
            |                   |                           \-----------J.Identifier | "line"
            |                   \---J.Try.Catch
            |                       |---J.ControlParentheses | "(Exception e)"
            |                       |   \-------J.VariableDeclarations | "Exception e"
            |                       |           |---J.Identifier | "Exception"
            |                       |           \-------J.VariableDeclarations.NamedVariable | "e"
            |                       |                   \---J.Identifier | "e"
            |                       \---J.Block
            |                           \-------J.MethodInvocation | "error("Exception", e)"
            |                                   |---J.Identifier | "error"
            |                                   \-----------J.Literal
            |                                       \-------J.Identifier | "e"
            \-------J.MethodDeclaration | "MethodDeclaration{T{name=error,return=void,parameters=[java.lang.String,java.lang.Exception]}}"
                    |---J.Modifier | "private"
                    |---J.Modifier | "static"
                    |---J.Primitive | "void"
                    |---J.Identifier | "error"
                    |-----------J.VariableDeclarations | "String s"
                    |   |       |---J.Identifier | "String"
                    |   |       \-------J.VariableDeclarations.NamedVariable | "s"
                    |   |               \---J.Identifier | "s"
                    |   \-------J.VariableDeclarations | "Exception e"
                    |           |---J.Identifier | "Exception"
                    |           \-------J.VariableDeclarations.NamedVariable | "e"
                    |                   \---J.Identifier | "e"
                    \---J.Block
------visitCompilationUnit LST Tree End------
```

### CODE REVIEW

File: rewrite-java/src/main/java/org/openrewrite/java/TreeVisitingPrinter.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+import static java.util.stream.StreamSupport.*;
+
+
+/**
+ * A visitor to print a tree visiting order and hierarchy in format like below.
+ * <pre>
+ * #root
+ * \---J.CompilationUnit | "..."
+ *     |---#JRightPadded
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/TreeVisitingPrinter.java:38]: Padding elements like `JRightPadded` are very common throughout the tree. True to the name, they keep track of whitespace may appear to the right of the AST element they enclose. For getting a sense of the structure of the tree these can be skipped over.

Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/TreeVisitingPrinter.java:38]: It makes sense to me, by removing the `JRightPadded` and `JLeftPadded` from the output, the example output tree print will reduce from 107 lines to 70 lines, let me make an update to skip over `JRightPadded` and `JLeftPadded` in the output.

Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/TreeVisitingPrinter.java]: skipped unvisited elements like `JRightPadded` or `JLeftPadded` from the output, but keeps an option to print all elements which may be extendable in the future.

**Comment by jkschneider** [comment]
Nice!

**Pull request merged**